Mathlib/LinearAlgebra/Matrix/Irreducible/Defs.lean

```diff
@@ -0,0 +1,253 @@
+      (Seneta, p.14).
+*   `IsIrreducible.transpose`: Shows that the irreducibility property is preserved under
+    transposition.
+
+## Implementation notes
+
+Throughout we work over a `LinearOrderedRing R`. Some results require stronger assumptions,
+like `PosMulStrictMono R` or `Nontrivial R`. Some statements expand matrix powers and thus require
+`[DecidableEq n]` to reason about finite sums.
+
+## References
+
+*   [E. Seneta, *Non-negative Matrices and Markov Chains*][seneta2006]
+
+## Tags
+
+matrix, nonnegative, positive, power, quiver, graph, irreducible, primitive, perron-frobenius
+-/
+namespace Matrix
+
+open Finset Quiver Quiver.Path
+
+variable {n R : Type*} [Fintype n] [Ring R] [LinearOrder R]
+
+/-- The directed graph (quiver) associated with a matrix `A`,
+with an edge `i ⟶ j` iff `0 < A i j`. -/
+def toQuiver (A : Matrix n n R) : Quiver n :=
+  ⟨fun i j => 0 < A i j⟩
+
+/-- A matrix `A` is irreducible if it is entrywise nonnegative and
+its quiver of positive entries (`toQuiver A`) is strongly connected. -/
+def IsIrreducible (A : Matrix n n R) : Prop :=
+  (∀ i j, 0 ≤ A i j) ∧ @IsSStronglyConnected n (toQuiver A)
+
+/-- A matrix `A` is primitive if it is entrywise nonnegative
+and some positive power has all entries strictly positive. -/
+def IsPrimitive [DecidableEq n] (A : Matrix n n R) : Prop :=
+  (∀ i j, 0 ≤ A i j) ∧ ∃ k > 0, ∀ i j, 0 < (A ^ k) i j
+
+variable {A : Matrix n n R}
+
+/-- If `A` is irreducible and `n` is non-trivial then every row has a positive entry. -/
+lemma IsIrreducible.exists_pos [Nontrivial n]
+    (h_irr : IsIrreducible A) (i : n) :
+    ∃ j, 0 < A i j := by
+  letI : Quiver n := toQuiver A
+  by_contra h_row
+  have no_out : ∀ j : n, IsEmpty (i ⟶ j) :=
+    fun j => ⟨fun e => h_row ⟨j, e⟩⟩
+  obtain ⟨j, hij⟩ := exists_pair_ne n
+  obtain ⟨p, hp_pos⟩ := h_irr.2 i j
+  have h_le : 1 ≤ p.length := Nat.succ_le_of_lt hp_pos
+  have ⟨v, p₁, p₂, _hp_eq, hp₁_len⟩ := p.exists_eq_comp_of_le_length (n := 1) h_le
+  have hlen_ne : p₁.length ≠ 0 := by simp [hp₁_len]
+  obtain ⟨c, p', e, rfl⟩ :=
+    (Quiver.Path.length_ne_zero_iff_eq_cons (p := p₁)).1 hlen_ne
+  have hp'_succ : p'.length + 1 = 1 := by
+    simpa [Quiver.Path.length_cons] using hp₁_len
+  have hp'0 : p'.length = 0 := by
+    exact Nat.succ.inj hp'_succ
+  have hi_c : i = c := Quiver.Path.eq_of_length_zero p' hp'0
+  subst hi_c
+  exact (no_out _).false e
+
+/--
+For a matrix `A` with nonnegative entries, the `(i, j)`-entry of the `k`-th power `A ^ k`
+is strictly positive if and only if there exists a path of length `k` from `i` to `j` in the
+quiver associated to `A` via `toQuiver`. -/
+theorem pow_entry_pos_iff_exists_path
+    [IsOrderedRing R] [PosMulStrictMono R] [Nontrivial R] [DecidableEq n]
+    (hA : ∀ i j, 0 ≤ A i j) (k : ℕ) (i j : n) :
+    letI := toQuiver A; 0 < (A ^ k) i j ↔ Nonempty {p : Path i j // p.length = k} := by
+  induction k generalizing i j with
```

Comment:
If we add this instance here then you don't need to keep adding it inside each subgoal below
```suggestion
  letI : Quiver n := A.toQuiver
  induction k generalizing i j with
```